plugin/pluginenv/environment.go

```diff
@@ -0,0 +1,123 @@
+// Package pluginenv provides high level functionality for discovering and launching plugins.
+package pluginenv
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/platform/plugin"
+)
+
+type APIProviderFunc func(*plugin.Manifest) (plugin.API, error)
+type SupervisorProviderFunc func(*plugin.BundleInfo) (plugin.Supervisor, error)
+
+// Environment represents an environment that plugins are discovered and launched in.
+type Environment struct {
+	searchPath         string
+	apiProvider        APIProviderFunc
+	supervisorProvider SupervisorProviderFunc
+	activePlugins      map[string]plugin.Supervisor
+}
+
+type Option func(*Environment)
+
+// Creates a new environment. At a minimum, the APIProvider and SearchPath options are required.
+func New(options ...Option) (*Environment, error) {
+	env := &Environment{
+		activePlugins: make(map[string]plugin.Supervisor),
+	}
+	for _, opt := range options {
+		opt(env)
+	}
+	if env.supervisorProvider == nil {
+		env.supervisorProvider = DefaultSupervisorProvider
+	}
+	if env.searchPath == "" {
+		return nil, fmt.Errorf("a search path must be provided")
```

Comment:
I guess I'm alright with that since this is isolated into it's own package but to me it seems like extra work on the caller when they could just return immediately in most cases if an AppError was returned already